### ISSUE TITLE
Performance issues on windows.

### ISSUE THREAD

**Issue by robfordww** [BLAKE3-team/BLAKE3#178]
I blake3ed a 64Gb file on a compact Asus system with 32 GB mem an nvme disks. It took approx 30 seconds. Approx 8 times faster than sha256. Then, the same file on windows 10 with even better hardware amd Ryzen 7 5800x, and faster nvme disks, just collapsed. It was more like 1m 30s now. 3 times as slow. Seems like the windows version got some performance issues?

**Comment by oconnor663** [comment]
It's likely this issue: https://github.com/BLAKE3-team/BLAKE3/issues/31. Could you see if a simple workaround like `b3sum < myfile` improves performance?

If this is the problem, what's happening is that `b3sum` likes to have many threads read from the file at many different offsets at the same time. If the file is cached in RAM, or on an SSD, this is usually a good strategy. But if the file is on a spinning disk, and too large to fit in RAM, then this is actually a really bad strategy. We end up blocking on IO, and doing lots of reads all over that thrash the disk, and it would've been a lot faster to just read the file from the front "the slow way".

We'll eventually work around this with a more complicated threading strategy, but in the meantime you might need to resort to explicit workarounds like `b3sum < myfile` or `b3sum myfile --num-threads=1`.

**Comment by robfordww** [comment]
Perhaps related to the spinning disk problem, but these machines both have latest grade NVME disks (faster than SSDs in general). These disks push 2-5 GB/s on sequential reads, like large files.

**Comment by robfordww** [comment]
I follow up with some more tests:
b3sum_windows_x64_bin.exe < bigfile68GB.dat   --> 45s on "warm cache"
b3sum_windows_x64_bin.exe bigfile68GB.dat --> 41s on "cold cache"
b3sum_windows_x64_bin.exe bigfile68GB.dat --> 40s on "warmcache"

**Comment by oconnor663** [comment]
What's the command that gives the 1m30s result? I would've exected it to be that second one (default command, cold cache), but it didn't repro there?

**Comment by robfordww** [comment]
@oconnor663   You are right!  The 1m 30s was for a cold cache run.  However, I think that I might have been somewhat fooled by the SSDs speed at that moment, as I can't recreate it.  I suspect the SSD might be somewhat slower after a huge fast file transfer, but I am not sure.  My case is weaker, but it is still there:  What I can replicate is 40s for 68 GB on windows, which is not bad.  But on linux, with "on paper" worse hardware, I run the same at 19s (I just tested it again).   Sequential reads on windows seems to benchmark the SSD at 3.5GB/s so I guess it must be cpu/memory bound perhaps?

**Comment by oconnor663** [comment]
I believe Windows is less aggressive than Linux about caching very large files in memory. So it could be that your Windows test is just reading more from disk than your Linux test is. But I don't have any concrete figures about that, just scattered experiences.

3.5 GB/s is gonna be in the ballpark of the CPU throughput of a single core (I get about 3.3 GB/s hashing from RAM on this laptop's i9-9880H), so if random reads reduce that throughput number, that could definitely explain some moderate performance effects here.

**Comment by sneves** [comment]
I wonder if increasing the read buffer from 64KB to something larger, say, 8MB, would make a difference.

**Comment by oconnor663** [comment]
@robfordww if you wanted to experiment with that, the read buffer size is [hardcoded here](https://github.com/BLAKE3-team/BLAKE3/blob/0.3.8/b3sum/src/main.rs#L261). One thing to keep in mind here is that this codepath is only taken when memory mapping is disabled, which is not the default. Hashing from stdin with `<` will disable memory mapping, or you can disable it with `--no-mmap`. (But `--num-threads=1` will not disable memory mapping by itself.)

**Comment by robfordww** [comment]
I got some new information to add:

I tried to do the 8MB read buffer thing, turns out that Windows(?) don't like that big buffers on the stack, so it crashed.  I also tried allocating the buffer on the HEAP,  but I didn't find too much difference.  It runs at approximately 35 seconds now for this 68GB file, on a cold boot.

I tried to run ubuntu 21.04 on an USB with the same hardware, and mount the NTFS from there by FUSE (ntfs-3g), downloaded the latest b3sum and ran.  Results was approximately 25 seconds of runtime.  I was not totally happy, and suspected the FUSE filesystem to be the bottleneck. 

I partitioned 110GB free space, did a real install of Ubuntu 21, and tried again.  Cold run time fell to amazing 13-14 seconds for a 68GB file!  Now we are talking speed :) , really impressive hashing.

![image](https://user-images.githubusercontent.com/24738409/122014702-b855a400-cdbf-11eb-9923-1d88860d7452.png)

Why is the difference so big between windows and linux on the same hardware?  I really cant tell.  But regardless, this is a really fast hashing algo. Well done.
